From fc565713eaa3d8e9eeff0596dc9227ea07d4364e Mon Sep 17 00:00:00 2001
From: Ryan Blue <blue@apache.org>
Date: Fri, 11 Sep 2015 10:31:38 -0700
Subject: [PATCH 166/176] PARQUET-335: Remove Avro check for MAP_KEY_VALUE.

This is not required by the map type spec. This does not affect data
written by the Avro object model because this bug is in the conversion
from a Parquet schema to an Avro schema. Files written with parquet-avro
do not convert the underlying schema because they use the Avro schema.

Author: Ryan Blue <blue@apache.org>

Closes #241 from rdblue/PARQUET-335-remove-key-value-check and squashes the following commits:

1fd9541 [Ryan Blue] PARQUET-335: Test that MAP_KEY_VALUE is not required.
247cc76 [Ryan Blue] PARQUET-335: Remove Avro check for MAP_KEY_VALUE.
---
 .../java/parquet/avro/AvroSchemaConverter.java     |    1 -
 .../java/parquet/avro/TestAvroSchemaConverter.java |   20 ++++++++++++++++++++
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/parquet-avro/src/main/java/parquet/avro/AvroSchemaConverter.java b/parquet-avro/src/main/java/parquet/avro/AvroSchemaConverter.java
index 8a62502..d2ee401 100644
--- a/parquet-avro/src/main/java/parquet/avro/AvroSchemaConverter.java
+++ b/parquet-avro/src/main/java/parquet/avro/AvroSchemaConverter.java
@@ -303,7 +303,6 @@ public class AvroSchemaConverter {
             }
             GroupType mapKeyValType = parquetGroupType.getType(0).asGroupType();
             if (!mapKeyValType.isRepetition(Type.Repetition.REPEATED) ||
-                !mapKeyValType.getOriginalType().equals(OriginalType.MAP_KEY_VALUE) ||
                 mapKeyValType.getFieldCount()!=2) {
               throw new UnsupportedOperationException("Invalid map type " + parquetGroupType);
             }
diff --git a/parquet-avro/src/test/java/parquet/avro/TestAvroSchemaConverter.java b/parquet-avro/src/test/java/parquet/avro/TestAvroSchemaConverter.java
index 35189b5..9ece31a 100644
--- a/parquet-avro/src/test/java/parquet/avro/TestAvroSchemaConverter.java
+++ b/parquet-avro/src/test/java/parquet/avro/TestAvroSchemaConverter.java
@@ -21,6 +21,7 @@ package parquet.avro;
 import com.google.common.collect.Lists;
 import com.google.common.io.Resources;
 import java.util.Arrays;
+import java.util.Collections;
 import org.apache.avro.Schema;
 import org.apache.hadoop.conf.Configuration;
 import org.codehaus.jackson.node.NullNode;
@@ -392,6 +393,25 @@ public class TestAvroSchemaConverter {
         "}\n");
   }
 
+  @Test
+  public void testParquetMapWithoutMapKeyValueAnnotation() throws Exception {
+    Schema schema = Schema.createRecord("myrecord", null, null, false);
+    Schema map = Schema.createMap(Schema.create(Schema.Type.INT));
+    schema.setFields(Collections.singletonList(new Schema.Field("mymap", map, null, null)));
+    String parquetSchema =
+        "message myrecord {\n" +
+            "  required group mymap (MAP) {\n" +
+            "    repeated group map {\n" +
+            "      required binary key (UTF8);\n" +
+            "      required int32 value;\n" +
+            "    }\n" +
+            "  }\n" +
+            "}\n";
+
+    testParquetToAvroConversion(schema, parquetSchema);
+    testParquetToAvroConversion(NEW_BEHAVIOR, schema, parquetSchema);
+  }
+
   public static Schema optional(Schema original) {
     return Schema.createUnion(Lists.newArrayList(
         Schema.create(Schema.Type.NULL),
-- 
1.7.9.5

